### PR TITLE
[Improve][API & Zeta] Using connector custom serializer encode/decode states

### PR DESCRIPTION
**Pull request by hailin0**
## Purpose of this pull request

* Zeta: Using connector custom serializer encode/decode states

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/context/SeaTunnelSplitEnumeratorContext.java [hunk 1]
```diff
@@ -67,22 +69,26 @@ public void assignSplit(int subtaskIndex, List<SplitT> splits) {
             log.warn("No reader is obtained, skip this assign!");
             return;
         }
+
+        List<byte[]> splitBytes =
+                splits.stream()
+                        .map(split -> sneaky(() -> task.getSplitSerializer().serialize(split)))
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/context/SeaTunnelSplitEnumeratorContext.java:75]: Use connector SplitSerializer

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/flow/SinkFlowLifeCycle.java [hunk 2]
```diff
@@ -184,10 +186,14 @@ public void received(Record<?> record) {
...
-                                        new SinkPrepareCommitOperation(
+                                        new SinkPrepareCommitOperation<CommitInfoT>(
                                                 barrier,
                                                 committerTaskLocation,
-                                                SerializationUtils.serialize(commitInfoT)),
+                                                commitInfoSerializer.isPresent()
+                                                        ? commitInfoSerializer
+                                                                .get()
+                                                                .serialize(commitInfoT)
+                                                        : null),
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/flow/SinkFlowLifeCycle.java:196]: Use connector CommitInfoSerializer

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/flow/SourceFlowLifeCycle.java [hunk 3]
```diff
@@ -338,21 +336,17 @@ public void restoreState(List<ActionSubtaskState> actionStateList) throws Except
...
                         .collect(Collectors.toList());
         try {
             runningTask
                     .getExecutionContext()
                     .sendToMember(
-                            new RestoredSplitOperation(
-                                    enumeratorTaskLocation,
-                                    SerializationUtils.serialize(splits.toArray()),
-                                    indexID),
+                            new RestoredSplitOperation(enumeratorTaskLocation, splits, indexID),
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/flow/SourceFlowLifeCycle.java:349]: Use connector SplitSerializer

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/operation/sink/SinkPrepareCommitOperation.java [hunk 4]
```diff
@@ -73,15 +72,24 @@ public int getClassId() {
...
+        ClassLoader mainClassLoader = Thread.currentThread().getContextClassLoader();
+
         if (commitInfos != null) {
-            committerTask.receivedWriterCommitInfo(
-                    barrier.getId(), SerializationUtils.deserialize(commitInfos, classLoader));
+            CommitInfoT deserializeCommitInfo = null;
+            try {
+                Thread.currentThread().setContextClassLoader(taskClassLoader);
+                deserializeCommitInfo =
+                        committerTask.getCommitInfoSerializer().deserialize(commitInfos);
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/operation/sink/SinkPrepareCommitOperation.java:88]: Use connector SplitSerializer

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/operation/source/AssignSplitOperation.java [hunk 5]
```diff
@@ -56,13 +55,22 @@ public void run() throws Exception {
...
                                     .getClassLoader();
-                    Object[] o = SerializationUtils.deserialize(splits, classLoader);
-                    task.receivedSourceSplit(
-                            Arrays.stream(o).map(i -> (SplitT) i).collect(Collectors.toList()));
+                    ClassLoader mainClassLoader = Thread.currentThread().getContextClassLoader();
+                    List<SplitT> deserializeSplits = new ArrayList<>();
+                    try {
+                        Thread.currentThread().setContextClassLoader(taskClassLoader);
+                        for (byte[] split : this.splits) {
+                            deserializeSplits.add(task.getSplitSerializer().deserialize(split));
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/operation/source/AssignSplitOperation.java:67]: Use connector SplitSerializer

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/operation/source/RestoredSplitOperation.java [hunk 6]
```diff
@@ -82,27 +86,31 @@ public void run() throws Exception {
...
                             taskExecutionService
                                     .getExecutionContext(taskLocation.getTaskGroupLocation())
                                     .getClassLoader();
+                    ClassLoader mainClassLoader = Thread.currentThread().getContextClassLoader();
+
+                    List<SourceSplit> deserializeSplits = new ArrayList<>();
+                    try {
+                        Thread.currentThread().setContextClassLoader(taskClassLoader);
+                        for (byte[] split : splits) {
+                            deserializeSplits.add(task.getSplitSerializer().deserialize(split));
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/operation/source/RestoredSplitOperation.java:102]: Use connector SplitSerializer

**Comment by Hisoka-X** [comment]
LGTM, but I think we should test it in two zeta server, it will more safe.

**Comment by hailin0** [comment]
> LGTM, but I think we should test it in two zeta server, it will more safe.

@Hisoka-X 
Yes, I have verified this change on a multi-node cluster

**Review by Hisoka-X — APPROVED** [review]

**Review by ic4y — APPROVED** [review]

**Pull request merged**